File: ch01/1.15/main.go
package main

import "fmt"

func duple_recursive(n int8, x string, los []string) []string {
  if n == 0 {
    return los
  }

  return append(duple_recursive(n - 1, x, los), x)
}

func duple(n int8, x string) []string {
  return duple_recursive(n, x, make([]string, 0))
}

func main() {
  fmt.Println("duple(2 , `3`: )", duple(2 , "3"))
  fmt.Println("duple(4 , `ha ha`: )", duple(4 , "ha ha"))
  fmt.Println("duple(0 , `bla`: )", duple(0 , "blah"))
}

File: ch01/1.15/main.hs
import Data.Char

dupleRecursive :: Integer -> String -> [String] -> [String]
dupleRecursive 0 x los = los
dupleRecursive n x los = x : dupleRecursive (n - 1) x los

duple :: Integer -> String -> [String]
duple n x = dupleRecursive n x []

-- | 'main' runs the main program
main :: IO ()
main = do
  print ("duple(2 , `3`): " ++ show (duple 2 "3"))
  print ("duple(4 , `ha ha`): " ++ show (duple 4 "ha ha"))
  print ("duple(0 , `bla`): " ++ show (duple 0 "blah"))
